torch/_dynamo/variables/higher_order_ops.py

```diff
@@ -786,18 +787,19 @@ def create_wrapped_node(self, tx, args, kwargs):
         # See NOTE [HigherOrderOperator tracing design] for more details
         checkpoint = tx.copy_graphstate()
         graph_checkpoint = tx.output.graph
+        flattened_kwargs = [kwargs[key] for key in sorted(kwargs.keys())]
+
         (
             body_r,
             body_graph,
             body_lifted_freevars,
         ) = speculate_subgraph(
             tx,
-            args[0],
-            [
-                *args[1:],
-            ],
+            args[0],  # function
+            [*(args[1:] + flattened_kwargs)],
```

Comment:
I'm kind of confused about the design. What I would have expected is:
- speculate_subgraph accepts both args and kwargs
- speculate_subgraph validates that (args, kwargs) have supported types
- speculate_subgraph may choose to manually add the args (not kwargs!) as subgraph inputs. In this case, there should be no kwargs.
- speculate_subgraph invokes call_function with args and kwargs. The free variable lifting mechanism promotes relevant Tensors (and, in the future, SymNodes) to being inputs of the subgraph.

There are a couple of different cases that the current code looks a bit suspicious for that I think the design solves. We could also turn these into more test cases:
- what happens if body_fun's kwargs aren't in order (it looks like we're sorting the kwargs?)
- what happens if body_fun has kwarg-only arguments?

Reply:
> * what happens if body_fun has kwarg-only arguments?

We get `TypeError: too many positional arguments`, I don't think we can replace kwarg-only args with positional at all. With this in mind, does the first approach from the PR of adding kwarg as input to the graph and returning the mapping of the updated `kwarg` name (in graph) makes sense?

Ref commit : https://github.com/pytorch/pytorch/pull/104180/commits/29da5817345bd7a9455c3a86114764bfee9d6198#diff-7bd43c6b174845f3450be42c8460ed021d1b39a98510cbea5d4e7e3a0b9d8d4f

Reply:
Also, following code fails with current approach (because we sort the kwargs based on keys).

```python
def f(x, y):
    def fn(y, x):
        return (y * 2) + x

    return wrap(fn, y=y, x=x)

x = torch.ones(()) * 2
y = torch.ones(()) * 8

a = f(x, y)
opt = torch.compile(f, backend="eager", fullgraph=True)
e = opt(x, y)

print(a, e)  # 18, 12
```

Reply:
> With this in mind, does the first approach from the PR of adding kwarg as input to the graph and returning the mapping of the updated kwarg name (in graph) makes sense?

We don't need to manually add the kwarg as an input to the subgraph graph, right? The proposal is that speculate_subgraph will directly pass the kwarg (and all args) to the body_function; when the kwarg gets used (like in a Tensor operation), it will manually get lifted into a positional input to the subgraph. 

So if we have a test case like the following:
```py
def f(*, x):
    return x.sin()

@torch.compile(backend='aot_eager')
def g(x):
    return wrap(f, x=x)

g(torch.randn(1))
```
the graph we get out should look like:
```
def forward(self, x):
    wrap(self.body_fn, x)

def body_fn(self, x):
    return x.sin()
```

Reply:
Something like the following? https://github.com/pytorch/pytorch/compare/rzou/kwargs_lift?expand=1

Reply:
Ah, got it. Sorry I got confused. Thanks!